ocaml-lsp-server/src/diagnostics.ml

```diff
@@ -0,0 +1,64 @@
+open Import
+
+module Uri = struct
+  include Uri
+
+  let compare x y = Ordering.of_int (Uri.compare x y)
+end
+
+module Uri_c = Comparable.Make (Uri)
+module Uri_set = Uri_c.Set
+
+type t =
+  { dune : (int, Uri.t * Diagnostic.t) Table.t
+  ; merlin : (Uri.t, Diagnostic.t list) Table.t
```

Comment:
Shouldn't dune provide the merlin diagnostics as well?

Reply:
How can dune provide the merlin diagnostics? Dune doesn't know anything about merlin.

We need to separate the fields because the diagnostics are updated differently for merlin and dune:

* For merlin, the diagnostics are always calculated for every document. Therefore, we always update diagnostics per document in bulk.
* For dune, the diagnostics can come in any order with add/remove events. Therefore, we keep unique id's for diagnostics to support efficient remove.

I will probably implement diagnostic id's on the client side to make this whole thing more obvious.

Reply:
> How can dune provide the merlin diagnostics? Dune doesn't know anything about merlin.

Dune at least generates the merlin files. For some reason, I assumed that it would also provide merlin diagnostics through an RPC request.

> For merlin, the diagnostics are always calculated for every document.

Are Dune events always global? If so why do we need the `Uri.t` for Dune as well? (trying to understand a bit better, I clearly have some gaps in my understanding of how diagnostics are generated)

Reply:
> Are Dune events always global? If so why do we need the Uri.t for Dune as well? (trying to understand a bit better, I clearly have some gaps in my understanding of how diagnostics are generated)

LSP requires every diagnostic to be associated to a file. That's why we need an associated Uri. Dune events are global but we can always find a suitable location that somewhat corresponds to the "source" of the error. This is what we're going to use to compute the uri.